pkg/logging/rate_limited_logger.go

```diff
@@ -0,0 +1,73 @@
+package logging
+
+import (
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"golang.org/x/time/rate"
+)
+
+// RateLimitedLogger wraps a zap.SugaredLogger that supports rate limiting.
+type RateLimitedLogger struct {
+	*Logger
+	rateLimiters *lru.Cache
+	// needs a mutex to protect burst count increase? //@TODO: khushboo
+}
+
+// NewRateLimitLogger returns a rate limited logger
+func NewRateLimitLogger(l *Logger, size int) *RateLimitedLogger {
+	cache, err := lru.New(size)
+	if err != nil {
+		l.Errorf("unable to create rate limiter cache for logger in module %q: %q", l.module.name, err)
+		return nil
+	}
+	return &RateLimitedLogger{
+		l,
+		cache,
+	}
+}
+
+// SetLimiter sets how many logLines are emitted in a given interval for a specific identifier.
+func (rl *RateLimitedLogger) SetLimiter(limiter string, logLines float64, interval time.Duration, burst int) {
+	if logLines == float64(burst) {
+		burst++
+	}
+	// Check first, and only add the limiter if it doesn't exist already.
+	_, _ = rl.rateLimiters.ContainsOrAdd(limiter, rate.NewLimiter(rate.Limit(logLines/interval.Seconds()), burst))
+}
+
+// ErrorL logs a templated error message if allowed by the rate limiter corresponding to the identifier
+func (rl *RateLimitedLogger) ErrorL(limiter string, template string, args ...interface{}) {
+	if rl.allowLog(limiter) {
```

Comment:
What do you think about removing `limiter` from the interface of `RateLimitedLogger` and calculating it on the fly via `fmt.Sprintf(template, args...)`? This will guarantee there is a rate limiter for every log entry, ensure different errors from the same hostname are not dropped, and simplify the interface.

Reply:
1) We decided in our conversation to rate limit on hostname so hence the limiter is hostname
2) As for different errors from the same host name - only auth failures are being rate limited. Any other error is not. So they will continue to be logged as is.